### ISSUE TITLE
Media Streaming Test

### ISSUE THREAD

**Issue by clarkzjw** [private-octopus/picoquic#1406]
Hi, I noticed that [`Media tests`](https://github.com/private-octopus/picoquic/blob/master/picoquictest/mediatest.c) that simulate media transmission were added recently.

I also started testing video transmission using picoquic in my own project recently. I followed the examples in `picoquicdemo` and `picoquic_sample` and tried to simulate the video streaming process. Here is my understanding regarding how to use picoquic during a video streaming session, correct me if I'm wrong. 

On the server side, I just use `picoquicdemo` or `picoquic_sample` to serve a directory containing video segments, each lasting 2 secs. And on the client side, the client first downloads the `.mpd` file and then downloads each video segment accordingly (regarding specific quality adaptation logic). 

Following the example in [`picoquic_sample`](https://github.com/private-octopus/picoquic/blob/master/sample/sample_client.c#L433-L445), I assume I can use a global QUIC context during each session. And maybe I can also reuse `one` client context (`sample_client_ctx_t client_ctx`) and client connection (`picoquic_cnx_t* cnx`) and a fixed number of streams for downloading different video segments? Does that make sense?

However, I'm not quite sure whether I'm on the right track. Currently, I made some quick attempts on https://github.com/clarkzjw/picoquic/commit/bce0c548fb2c11025fd5c1a5340185e9cf832f34. When downloading the second file, the error shows `Error 1029, cannot initialize stream for file number 0` which seems like cannot find available streams.

**Comment by huitema** [comment]
What you describe is pretty similar to MPEG DASH. That works, as long as you are OK with latency measured in seconds. Also, you have to be careful to start segments at a "natural" boundary in the video, such as an I frame.

You can look at https://github.com/Quicr/quicrq/ for two different approaches, one mapping a video stream directly to a QUIC stream, and the other sending the video as a series of datagrams. Datagrams result in somewhat better handling of relays. In both case, the key is to avoid queues by building some kind of "selective drop" logic at the sender, i.e., dropping low priority data if congestion is noticed.

**Comment by clarkzjw** [comment]
> What you describe is pretty similar to MPEG DASH. That works, as long as you are OK with latency measured in seconds.

Yes, indeed. My project will focus on scheduling algorithms. 

Thanks, I'll take a look at quicrq.

**Comment by clarkzjw** [comment]
The `picoquic_sample` creates a number of streams according to the number of files. https://github.com/private-octopus/picoquic/blob/master/sample/sample_client.c#L544-L549

```C
        /* Create a stream context for all the files that should be downloaded */
        for (int i = 0; ret == 0 && i < client_ctx.nb_files; i++) {
            ret = sample_client_create_stream(cnx, &client_ctx, i);
            if (ret < 0) {
                fprintf(stderr, "Could not initiate stream for fi\n");
            }
        }
```
If I would like to reuse existing connections and streams to download files, for example, creating a fixed number of streams within one quic connection and reusing them repeatedly, what should I pay attention to?

**Comment by huitema** [comment]
Reusing streams for multiple files? I wonder why you would want to do that. QUIC is very much centered on having many streams, it is not a scarce resource, and it is not expensive. Transaction systems like DNS over QUIC typically use one stream per transaction.

But opportunity questions apart, a given QUIC stream can behave as a TCP connection, i.e., a reliable stream of bytes. Your application code can decide to send on a stream a series of well delineated messages. For example, HTTP 3 will send on a stream a series of frames, see [section 4 of RFC 9114](https://www.rfc-editor.org/rfc/rfc9114.html#name-http-message-framing). Typically, that would be a series of data segments preceded by a length field.

**Comment by clarkzjw** [comment]
Thanks for the reply. I think I might misunderstand the stream concept before.
 
Also, per Section 4 of [Applicability of the QUIC Transport Protocol](https://quicwg.org/ops-drafts/draft-ietf-quic-applicability.html#name-use-of-streams), 

> A single stream provides ordering. If the application requires certain data to be received in order, that data should be sent on the same stream. There is no guarantee of transmission, reception, or delivery order across streams.

> Multiple streams provide concurrency. Data that can be processed independently, and therefore would suffer from head of line blocking if forced to be received in order, should be transmitted over separate streams.

So it is more reasonable to reuse the same connection context `picoquic_cnx_t` and multiple streams during a session of multiple file downloads, right?

**Comment by huitema** [comment]
Yes, that is what most applications sending files do. Files are defined as ordered sequences of bytes, just like streams, so mapping one on the other is most natural. 

But we were discussing media streaming earlier. Media streams are not files. They are structured as sequences of media frames, sometimes with additional structure like the frame types I, B and P of MPEG. I have observed various projects testing at least 4 different approaches:

* One QUIC stream per media stream:
    - plus: simple.
    - minus: subject to head of line blocking. A media transmission error will cause a small delay until the correction is received. This can be mitigated by buffering, at the cost of latency.
    - minus: no easy way to support fast forward or roll back, apart from downloading the whole file.
* Break the big media file in a series of short files. Typically the list of short files is collected in a "manifest" that the client downloads at the beginning of the session. The client then downloads the short files one by one as the playback progresses. This is what MPEG DASH and similar solutions do.
    - plus: easier to support fast forward or roll back.
    - plus: can easily plug in some form of rate adaptation by providing several versions of short files encoded at different rates. The client then chooses the adequate version based on download and playback experience.
    - minus: still subject to head of line blocking.
* Map each media frame to a separate QUIC stream. This implemented by Twitch in their WARP project.
   - plus: can avoid head of line blocking. If one frame is delayed by retransmissions, it can be skipped if the next one is already there.
   - minus: more complex to implement. Each frame will need a header specifying the media stream and the order of the frame in that stream, which creates some overhead.
* Send the media stream as a series of QUIC datagram frames. This is done for example with "RTP over QUIC", or with QUICR.
   - plus: maps very well to real time requirements, avoids head of line blocking.
   - minus: applications have to deal with a lot of complexity.
   - minus: requires support of the Datagram extension in the QUIC stacks.

**Comment by clarkzjw** [comment]
> I have observed various projects testing at least 4 different approaches

Yes, my work as of right now falls into the second category. 

Thanks and I'll update here when I have more info to share or when I encounter another problem.

**Comment by clarkzjw** [comment]
Hi, recently, based on the original `picoquicdemo`, I modified it a little bit to let the client reuse the same QUIC connection to download multiple files (e.g., download multiple video segments similar to DASH), i.e.,

1. The client creates a QUIC connection
2. For each file download, a new stream is created

Note that, in the original `client_loop_cb`, in the case of `picoquic_packet_loop_after_receive`, the connection is closed if there are no more open streams. What I did is that instead of `picoquic_close(cb_ctx->cnx_client, 0);`, the client is reinserted by wake time

```C
if (cb_ctx->demo_callback_ctx->nb_open_streams == 0) {
    cb_ctx->demo_callback_ctx->connection_closed = 1;
    cb_ctx->cnx_client->cnx_state = picoquic_state_disconnected;
    picoquic_reinsert_by_wake_time(cb_ctx->cnx_client->quic, cb_ctx->cnx_client, picoquic_get_quic_time(cb_ctx->cnx_client->quic));
}
```

And when downloading newer files, the connection state is reset to `picoquic_state_ready` and new streams are created, e.g.,

```C
callback_ctx->connection_closed = 0;
callback_ctx->connection_ready = 1;
cnx_client->cnx_state = picoquic_state_ready;

for (int i = 0; i < callback_ctx->nb_demo_streams; i++) {
    callback_ctx->demo_stream[i].stream_id = stream_id + i * 4;
    ret = demo_client_start_streams(cnx_client, callback_ctx, callback_ctx->demo_stream[i].stream_id);
    if (ret != 0) {
        printf("start stream failed: %d\n", ret);
        return 0;
    }
}
```

Currently, the behavior of my codes shows that, after a certain amount of time (or after a certain amount of streams created?), the packets sent from the server get lost, e.g.,

<img width="1396" alt="Snipaste_2022-10-06_16-06-49" src="https://user-images.githubusercontent.com/2663319/194436058-35204135-e6bb-4797-ad69-9b29cc6995bb.png">

And all the subsequent packets are lost.

Any possible ideas about how to debug this problem or where should I pay attention to? Thanks.

The qlog files are attached.
[2e78b7a38235fc2e.1709.server.qlog.txt](https://github.com/private-octopus/picoquic/files/9729186/2e78b7a38235fc2e.1709.server.qlog.txt)
[2e78b7a38235fc2e.client.qlog.txt](https://github.com/private-octopus/picoquic/files/9729187/2e78b7a38235fc2e.client.qlog.txt)

**Comment by huitema** [comment]
It seems that at some point the client stops receiving data. This kind of failure can happen for many reasons, but most probably the client is not taking time to read the socket - some bug in the receive loop. You could verify that using wireshark, see whether server packets are arriving at the client.

Note that if the client is state "disconnected", it will indeed not process any incoming packet. Reinserting by wait time just means that the client will immediately try to send something. If you want to continue the connection, you should not mess with its state. Instead, you should maybe try download another file. Why are you doing that?

**Comment by clarkzjw** [comment]
> It seems that at some point the client stops receiving data. This kind of failure can happen for many reasons, but most probably the client is not taking time to read the socket - some bug in the receive loop. You could verify that using wireshark, see whether server packets are arriving at the client.

Thanks for the advice, I'll try to look into that.

> Note that if the client is state "disconnected", it will indeed not process any incoming packet. Reinserting by wait time just means that the client will immediately try to send something. If you want to continue the connection, you should not mess with its state. Instead, you should maybe try download another file. Why are you doing that?

Mostly because of the adaptive bitrate decision during the playback, after the current video segment file is downloaded, the playback buffer could be full and the client may need to wait a uncertain amount of time before begin downloading the next video segment. So I thought it might be easier to exit the current packet loop first. And intuitively, I saw `picoquic_packet_loop` exits if the callback function returns `PICOQUIC_NO_ERROR_TERMINATE_PACKET_LOOP`, so I made the connection to be disconnected first.

**Comment by huitema** [comment]
You should not try to manipulate directly the state variable. Instead, you probably need to write your own application. It could well be derived from the picoquicdemo code, but a lot of the code will have to change. The picoquic demo code is driven by scenarios, in which the client downloads a series of files and then exits. From what you told me, it seems that you want to download a long series of media segments, with some logic to have P downloads in parallel, and start downloading segment N+P as soon as segment N has finished playing. You should write your application loop to do that.

**Comment by clarkzjw** [comment]
> it seems that you want to download a long series of media segments, with some logic to have P downloads in parallel, and start downloading segment N+P as soon as segment N has finished playing

Yes indeed. Thanks for the previous advice and I'll look into that.

**Comment by clarkzjw** [comment]
> It seems that at some point the client stops receiving data. This kind of failure can happen for many reasons, but most probably the client is not taking time to read the socket - some bug in the receive loop. You could verify that using wireshark, see whether server packets are arriving at the client.

I tried to look into this issue, and it seems that, after a certain amount of time, on the client side, the socket select always returns 0, https://github.com/private-octopus/picoquic/blob/master/picoquic/picosocks.c#L1179

```C
ret_select = select(sockmax + 1, &readfds, NULL, NULL, &tv);
```

which means the timeval `tv`, somewhat determined by `delta_t = picoquic_get_next_wake_delay(quic, current_time, delay_max);` that expires. 

In previous demos to test QUIC connection reuse, I just downloaded multiple files one after another, without any delays or wait in between. Right now, if I introduce a small delay(e.g., 0.1s) between each download, the above issues disappear. But I haven't found the root cause yet.

**Comment by huitema** [comment]
That means the timeval is set to zero. This could very well be a bug. The value is set in the send loop when executing the function `picoquic_prepare_packet_ex`. There are many tests, for various conditions of the type "trying to send a packet or a frame of type X, but cannot do it right now". Each of these tests may reset the "next wake time". If a test is wrong, it may reset the next wait tie to the current time, but when ask later to send a packet find that it is not yet ready, and reset the time again. That would be a bug. The good news is that when a sender function does that, it calls the macro `SET_LAST_WAKE`, which records the line of code that reset the wake time in variables `(quic)->wake_file` and `(quic)->wake_line`.

First debug step could be: run your program, with a test to check whether select returns 0 several times in a row without sending any packet. If that happens, find the values of `(quic)->wake_file` and `(quic)->wake_line`. We could debug it from there.

**Comment by clarkzjw** [comment]
Thanks for the info! I'll update when I have more debug info available.

**Comment by clarkzjw** [comment]
> First debug step could be: run your program, with a test to check whether select returns 0 several times in a row without sending any packet. If that happens, find the values of (quic)->wake_file and (quic)->wake_line. We could debug it from there.

When `ret_select = 0` in [`ret_select = select(sockmax + 1, &readfds, NULL, NULL, &tv);`](https://github.com/private-octopus/picoquic/blob/master/picoquic/picosocks.c#L1179), thus [`bytes_recv == 0`](https://github.com/private-octopus/picoquic/blob/master/picoquic/sockloop.c#L289), I printed out `quic->wake_file` and `quic->wake_line`. 

And it turns out that when `select returns 0 several times in a row without sending any packet`, `SET_LAST_WAKE` is called in `picoquic_retransmit_needed_packet` in `sender.c`, https://github.com/private-octopus/picoquic/blob/master/picoquic/sender.c#L1569. 

```C
    if (should_retransmit == 0) {
        /*
         * Always retransmit in order. If not this one, then nothing.
         * But make an exception for 0-RTT packets.
         */
        if (old_p->ptype == picoquic_packet_0rtt_protected) {
            *continue_next = 1;
        }
        else {
            if (next_retransmit_time < *next_wake_time) {
                *next_wake_time = next_retransmit_time;
                SET_LAST_WAKE(cnx->quic, PICOQUIC_SENDER);
            }
            /* Will not continue */
            *continue_next = 0;
        }
    }
```

And `old_p->ptype` seems to be `picoquic_packet_1rtt_protected`. In the example qlog attached, `1RTT:235`-`1RTT:237` were received on the server side, but the client thought they were lost. Thus although the server sends more data for the next stream, the client cannot receive them.

Is it possible it's related to [`picoquic_retransmit_needed_by_packet`](https://github.com/private-octopus/picoquic/blob/master/picoquic/sender.c#L1298-L1310)? 


[85c8904ade632f29.6c85.server.qlog.txt](https://github.com/private-octopus/picoquic/files/9759094/85c8904ade632f29.6c85.server.qlog.txt)
[85c8904ade632f29.client.qlog.txt](https://github.com/private-octopus/picoquic/files/9759095/85c8904ade632f29.client.qlog.txt)

**Comment by huitema** [comment]
I think there is a problem with the way you have modified `picoquicdemo`. The first set of losses is seen after stream 28 finishes. In the log, I see on the client side the message:
```
"message": "Out of 1 zero RTT packets, 1 were acked by the server."
```
That message is only produced in picoquicdemo.c, after the call to `picoquic_packet_loop` returns. I don't know what you did, but if the client exits the socket loop, it cannot just try to run it again. The old sockets are closed, the client will create new sockets, new port numbers. Lots of things are going to be wrong, including probably the handling of time. Indeed, no packet seems to be received by the client after that, probably because the server sends then to the old client socket instead of the new one.

You cannot implement your application by just installing a loop in `quic_client`. You have to replace the scenario logic in the client code. The scenario-driven logic is implemented in `democlient.c`. You will probably need to use a different logic, which probably means creating a different client, something like `mediaclient.c` that implements the logic of pulling media segments, assigning them priorities, etc.

**Comment by clarkzjw** [comment]
> The old sockets are closed, the client will create new sockets, new port numbers.

It's true. And I think I might find the problem here. I noticed a log on the server side:

```
"message": "Path 7 promoted to default at T=0.532930s, Local: 127.0.0.1:20753, Remote: 127.0.0.1:43053"
```

Each time the client creates a new pair of sockets and ports, it's considered as a new path, and thus the connection is migrated to the new one. Then it reaches `PICOQUIC_NB_PATH_TARGET`, thus the subsequent file downloading failed. 

I did replace some logic in `democlient.c`, and previously I thought it was easier to get started with, now I have to carefully reconsider and rewrite again. Thanks for the advice!

**Comment by huitema** [comment]
The test of PICOQUIC_NB_PATH_TARGET fails when the client tries to create too many paths too quickly. The logic is to consider only a certain number of paths at any given time. For example, we could see a DDOS attack in which the client purposely changes the source path on every new packet. If the server created a new path each time, then the memory would soon be exhausted. I think your code was testing that behavior.

We may want to revisit the way picoquic handle path creations. It may be that it does not clean up the list of old abandoned paths soon enough. On NAT rebinding, the server is supposed to send a challenge to both the new and the old path, and delete the old path if it does not respond to the challenge after a number of repetitions. Maybe that's too conservative. Maybe the old path could be deleted when the new one is confirmed. But this is a tricky area, because deleting too soon may also facilitate some attacks. We would need extensive testing.

**Comment by clarkzjw** [comment]
I see, then it makes sense why previously a slight delay between each download happened to work, though that's not the solution.

**Comment by clarkzjw** [comment]
> It may be that it does not clean up the list of old abandoned paths soon enough.

Where should I pay attention if I want to modify these logics? 

Or is it possible to temporarily workaround it by just modifying `PICOQUIC_CONNECTION_ID_MAX_SIZE`, `PICOQUIC_NB_PATH_TARGET`, `PICOQUIC_STORED_IP_MAX`? I tried to increase the values for these macros, but it seems it's also related to https://github.com/private-octopus/picoquic/blob/master/picoquic/quicctx.c#L2132-L2133.
```C
        if (nb_cid_received >= cnx->local_parameters.active_connection_id_limit + nb_cid_retired_before) {
            ret = PICOQUIC_TRANSPORT_CONNECTION_ID_LIMIT_ERROR;
        }
```

**Comment by huitema** [comment]
If you want quick and dirty fixes, you may just change the API to the socket loop to pass a socket ID -- as a read-write variable. If that is set to "INVALID_SOCKET" the socket loop would create a socket and document it. If not, it would reuse the valid socket. In both cases, not close it. That way, when you run your loop, the soocket and port will stay the same, and there will not be a migration.

**Comment by huitema** [comment]
@clarkzjw Is this issue still open for you?

**Comment by clarkzjw** [comment]
I'm closing this for the time being。
